### PR TITLE
Client metrics: draft of components API and interaction

### PR DESCRIPTION
**Pull request by maksymmalyhin**
Context b/171481539.

- `GDTCORClientMetricsController` API and related data classes
- required storage (`GDTCORStorageProtocol`) API updates
- `ClientMetrics` proto and generated code

### CODE REVIEW

File: GoogleDataTransport/GDTCCTLibrary/Protogen/nanopb/client_metrics.nanopb.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2019 Google
```
Comment (paulb777) [inline GoogleDataTransport/GDTCCTLibrary/Protogen/nanopb/client_metrics.nanopb.c:2]: 2021

File: GoogleDataTransport/GDTCCTLibrary/Protogen/nanopb/client_metrics.nanopb.h [hunk 2]
```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2019 Google
```
Comment (paulb777) [inline GoogleDataTransport/GDTCCTLibrary/Protogen/nanopb/client_metrics.nanopb.h:2]: 2021

File: GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORDroppedEventsCounter.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+/// Represents a number of events dropped for a particular mapping ID with a specified reason.
+/// The model is also used as a stored object for the persistent storage.
+@interface GDTCORDroppedEventsCounter : NSObject <NSSecureCoding>
+
+/// Number of events dropped.
+@property(nonatomic, readonly) NSUInteger eventCount;
+
+/// The reason why the events were dropped.
+@property(nonatomic, readonly) GDTCOREventDropReason dropReason;
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORDroppedEventsCounter.h:31]: Do all of the events in the eventCount have the same dropReason?

File: GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORClientMetricsControllerProtocol.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                             count:(NSUInteger)count;
+
+@end
+
+/// Client metrics controller API
+@protocol GDTCORClientMetricsControllerProtocol <GDTCORClientMetricsLogger>
+
+/// Retrieves metrics collected since the last metrics upload.
+/// NOTE: the metrics are not reset until `confirmSendingClientMetrics:` method is called. We assume
+/// that upload doesn't happen concurrently, so no data race is expected.
```
Comment (ncooke3) [inline GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORClientMetricsControllerProtocol.h:44]: More of a GDT question but how do we know that the upload doesn't happen concurrently?

File: GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h [hunk 5]
```diff
@@ -115,16 +126,21 @@ typedef void (^GDTCORStorageBatchBlock)(NSNumber *_Nullable newBatchID,
...
- * @param onComplete The callback that will be invoked once storage size calculation is complete.
- */
-- (void)storageSizeWithCallback:(void (^)(GDTCORStorageSizeBytes storageSize))onComplete;
-
 @end
 
 // TODO: Consider complete replacing block based API by promise API.
 
+@protocol GDTCORLibraryData <NSSecureCoding, NSObject>
+@end
```
Comment (ncooke3) [inline GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h:134]: Does/how does this type relate to `GDTCORClientMetrics`?

File: GoogleDataTransport/ProtoSupport/Protos/client_metrics.proto [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// Metrics per log source.
+message LogSourceMetrics {
+  // A LogSource uniquely identifies a logging configuration. log_source should
+  // contains a string value of the LogSource from
+  // google3/wireless/android/play/playlog/proto/clientanalytics.proto
+  string log_source = 1;
+
+  repeated LogEventDropped log_event_dropped = 2;
```
Comment (ncooke3) [inline GoogleDataTransport/ProtoSupport/Protos/client_metrics.proto:60]: Based on DD I'm referencing, is this supposed to have `num_data_events` and `num_telemetry_events` fields?

File: GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h (hunk 5 above)
Comment (ncooke3) [reply GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h:134]: Followup, is the `GDTCORClientMetrics` ultimately what we are storing?

**Review by ncooke3 — COMMENTED** [review]
Generally LGTM. Leaving several questions to help clarify my understanding– my GDT knowledge is lacking...

File: GoogleDataTransport/ProtoSupport/Protos/client_metrics.proto [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+message ClientMetrics {
+  // The window of time over which the metrics are evaluated.
+  TimeWindow window = 1;
+
+  repeated LogSourceMetrics log_source_metrics = 2;
+
+  GlobalMetrics global_metrics = 3;
+
+  // The bundle ID on Apple platforms (e.g., iOS) or the package name on Android
+  string app_namespace = 4;
```
Comment (ncooke3) [inline GoogleDataTransport/ProtoSupport/Protos/client_metrics.proto:23]: Based on DD I'm referencing, should this be `app_bundle`?

File: GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORClientMetricsControllerProtocol.h [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+/// @param mappingID The mappingID (log source).
+/// @param count The number of dropped events with specified reason and count.
+- (void)logEventsDroppedWithReason:(GDTCOREventDropReason)reason
+                         mappingID:(NSString *)mappingID
+                             count:(NSUInteger)count;
+
+@end
+
+/// Client metrics controller API
+@protocol GDTCORClientMetricsControllerProtocol <GDTCORClientMetricsLogger>
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORClientMetricsControllerProtocol.h:40]: It's an early idea and we may eventually merge these protocols into one. The main motivation is to separate APIs not directly related to each other, so the clients have access only the required API which should reduce probability of errors and simplify mocking for testing. But seems like we won't have a use case for that and we will merge these to a single protocol.

File: GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORClientMetricsControllerProtocol.h (hunk 4 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORClientMetricsControllerProtocol.h:44]: Currently we explicitly limit concurrent uploads with a serial queue, in the `GDTCCTUploader`, see more details [here](https://github.com/google/GoogleDataTransport/blob/7fb27ea49414b9c5483503cd06baa821c8654d1e/GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m#L71).

File: GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORDroppedEventsCounter.h (hunk 3 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCORDroppedEventsCounter.h:31]: Yes, the intention is to maintain as many counters as many reasons and mapping IDs we have.

File: GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCOREventDropReason.h [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+typedef NS_ENUM(NSInteger, GDTCOREventDropReason) {
+  GDTCOREventDropReasonUnknown = 0,
+  GDTCOREventDropReasonMessageTooOld,
+  GDTCOREventDropReasonStorageFull,
+  //  GDTCOREventDropReasonPayloadTooBig, ??
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/ClientMetrics/GDTCOREventDropReason.h:23]: Currently we don't limit the maximum payload size in Apple SDK, so this reason will never be used. That's why it's commented so far. Let me remove it to be reintroduced when relevant.

File: GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h (hunk 5 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h:134]: Good question. Yeah, maybe it makes sense to be less generic here and use more specific API until generic one is actually required. I'll add a TODO to reconsider it.

File: GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h [hunk 10]
```diff
@@ -152,6 +168,19 @@ typedef void (^GDTCORStorageBatchBlock)(NSNumber *_Nullable newBatchID,
...
+// Library data.
+
+/// Atomically retrieves and stores the library data for a given key. The library data is not
+/// counted toward the events size limit.
+/// @param key The key to get and update data for.
+/// @param readWriteBlock The block where the fetched data for the given key will be passed and to
+/// return the updated value to store. See also `GDTCORStorageLibraryDataReadWriteBlock`.
+/// @return A promise that is resolved with the updated value in the case of success and is rejected
+/// with an error otherwise.
+- (FBLPromise<id<NSSecureCoding>> *)
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h:180]: Yes, it should! Good catch.

File: GoogleDataTransport/ProtoSupport/Protos/client_metrics.proto (hunk 7 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/ProtoSupport/Protos/client_metrics.proto:38]: I think DD is slightly outdated. Currently the source of truth is the proto at google3/firebase/firelog/proto/gdt_client_analytics.proto which is used by the backend and copied to Apple and Android SDK.

File: GoogleDataTransport/ProtoSupport/Protos/client_metrics.proto (hunk 6 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/ProtoSupport/Protos/client_metrics.proto:75]: same as [above](https://github.com/google/GoogleDataTransport/pull/40/files#r717703604)

**Comment by maksymmalyhin** [comment]
I'm going to merge the PR to the feature branch to keep building the implementation on top. Feel free to leave comments to be addressed in the followup PRs.

File: GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h [hunk 11]
```diff
@@ -168,4 +199,21 @@ FOUNDATION_EXPORT
...
+/// The method is called when storage removes expired events.
+/// @param storage An instance of the storage.
+/// @param eventCount A number of events removed.
+- (void)storage:(id<GDTCORStoragePromiseProtocol>)storage
+    didRemoveExpiredEvents:(NSUInteger)eventCount;
+
+/// The method is called when an event passed to `storeEvent:` method was not stored due to reaching
+/// the maximum storage capacity.
+/// @param storage An instance of the storage.
+/// @param event The event that has beed dropped.
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/Internal/GDTCORStorageProtocol.h:214]: nit: s/beed/been/

File: GoogleDataTransport/GDTCORTests/Unit/Helpers/GDTCORClientMetricsControllerFake.h [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+/// corresponding arguments.
+@property(nonatomic, nullable, copy) void (^logEventsDroppedHandler)
+    (GDTCOREventDropReason reason, NSString *mappingID, NSUInteger count);
+
+/// Is called when `getMetrics` method is called. The returned promise will be returned from the
+/// method.
+@property(nonatomic, nullable, copy) FBLPromise<GDTCORClientMetrics *> * (^getMetricsHandler)(void);
+
+/// Is called when `confirmSendingClientMetrics:` method is called. The returned promise will be
+/// returned form the method.
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORTests/Unit/Helpers/GDTCORClientMetricsControllerFake.h:35]: s/form/from/

Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORTests/Unit/Helpers/GDTCORClientMetricsControllerFake.h:35]: Thanks! Fixed in mm/metrics2 branch.

**Pull request merged**